### ISSUE TITLE
Database cleanup does not work for Fanbox images

### ISSUE THREAD

**Issue by SimonPilkington** [Nandaka/PixivUtil2#690]
Presumably caused by #680.

**Comment by SimonPilkington** [comment]
Other database function may be broken as well. Please verify.

**Comment by bluerthanever** [comment]
Okay... Will have a look at it though. But could you give more detailed description?

**Comment by SimonPilkington** [comment]
https://github.com/Nandaka/PixivUtil2/blob/3b026ae5d5e1eb8de19cce0a994b7cd7c087a3af/PixivDBManager.py#L753 
The fanbox_master_post table you added in your PR is not checked.

**Comment by bluerthanever** [comment]
Oh yes.. Sorry, I had never used that function before...
Will be fixing right away.

And you mentioned `other` database functions? Do you mind telling me what they are?

**Comment by SimonPilkington** [comment]
Basically I meant it would be good to go through PixivDBManager.py and see if any other functions there are affected by your changes. I only confirmed cleanup and interactive cleanup no longer work as expected.

**Comment by bluerthanever** [comment]
Ah. I see. I had just looked at the `cleanup` function.... And actually I didn't know about this function. And I don't really use it so I didn't write a cleanup function for fanbox table.

And I have not touched other functions for regular tables, but I will double check.

May I ask how do you want the cleanup function to be? Same as the clean up function for regular images? Cuz it kinda was designed differently that it doesn't record download files, just some brief information, so it may be a little different... 
If necessary I can use the file formats...
But please do tell me how you want it to be. Thank you.

**Comment by SimonPilkington** [comment]
I am not the only user of this software, so my opinion alone should not define what happens, but for my part, I would expect the cleanup function to detect when something is missing from the file system so that it can be redownloaded if necessary. Beyond that, I would expect all database management functions to work for Fanbox posts as they did before. Breaking changes to UX should not happen.

@Nandaka, please weigh in on this issue too. @bluerthanever, it may be pertinent to wait for @Nandaka before you start any corrections.

**Comment by bluerthanever** [comment]
Okay... I understand... I will be waiting.

But before this, the program does not keep records of FANBOX download history. It just goes through all files the posts until the last page and skip downloading if the files exist...

It kinda does now, just in a really simple way. I will write it in the end of this post, if you are not interested....

But if it's just that you want to check the existence of each file every single time, I think a switch would do.... But I will wait.



How I implement this:
1. There's `published date` and `updated date` for each post, it's retrieved in json from API
2. When ever the post information is retrieved, its `published date` and `retrieved date` would be recorded in database
3. If the post is restricted, it will not be processed and the program goes to the next post.
4. If it's not restricted,  recorded `updated date` will be retrieved from database. And if it's empty or earlier than the new `updated date`, which means either the post is not downloaded before or that it's updated, the program would start processing it.
5.After processing it, the `updated date` would be written into database

And in case some files are like deleted or moved, that's why I added the delete FANBOX download history by post_id and member_id function...

**Comment by SimonPilkington** [comment]
I think it is useful to know with certainty which posts exist locally and which don't. Some artists move posts to higher tiers after a period elapses, which makes them impossible to redownload, so the sooner I can know that a post is missing from my file system, the better. As is, verifying manually seems to be the only option.

**Comment by bluerthanever** [comment]
So.... the request for cleaning up fanbox database table still stands right?
I will be waiting for my trial.... haha.

**Comment by SimonPilkington** [comment]
Yes, I would like to have it.

It is not a trial. We are discussing how to solve the issue. The requirements should be defined before you start working, and the owner of the repository should have a say in them.

**Comment by bluerthanever** [comment]
Haha. I am just joking.
It was pretty much based on my own requirement, for I am a user too, and nobody else was requesting, and also Nandaka-san himself doesn't use FANBOX....

**Comment by Nandaka** [comment]
Yep, I have no access to Fanbox as I don't really use it (free user here).

**Comment by SimonPilkington** [comment]
Since Nandaka has no strong opinions, I would suggest implementing working database cleanup for Fanbox posts as I said. This means detecting when files are missing on the file system and removing the corresponding entries from the database. It can be done either by checking the existence of each individual file or just the destination directory. I believe the latter would suffice.

**Comment by bluerthanever** [comment]
Sorry for the late reply.
Hmm... I think the latter might do but will be low efficient, and not accurate, given the circumstance, that the filename formats might be quite different for different users, and some elements might be missing (not recorded in db), and using regular expression to iterate each file and compare would be low efficient,
And the former option, would require a new table to record every file downloaded. This should be higher efficient, to sacrifice space for speed. But if the post contains no images or files, and the user saves the texts by image info and they delete files, this would not be detected at all...

If it's okay with you, I will probably go with the former one, this might take some time though.

**Comment by SimonPilkington** [comment]
If you prefer to record each file, fine. It will not significantly increase the size of the database anyway. Take your time with the development. I do not need this fixed immediately.

**Comment by bluerthanever** [comment]
Actually, I would not delete the posts I downloaded, or any images at all. so I don't really need to know which files do not exist locally, And I pretty much think posts which the author moves to higher tier will be downloaded when you run all posts through and hit it again, because the `updatedDate` would be changed.
But as you said, it's not my preference, but to satisfy the many users of this tool, And also I wasn't comprehensive enough.

**Comment by bluerthanever** [comment]
This should have been fixed in #695. Not sure if you have used it so far and whether it meets your expectation.
Please leave a feedback. Thanks.
